### PR TITLE
Refactor test_torch.py

### PR DESCRIPTION
**Pull request by johannesjmeyer**
New fixtures were introduced to declutter the code. Docstrings are now
in multiline format. Asserts use system-wide tolerance.

Issues: #99

### CODE REVIEW

**Comment by johannesjmeyer** [comment]
Do you whish that I also restructure the code a little? There are a lot of long tests in there.

**Comment by josh146** [comment]
@johannesjmeyer, yes, that would be a great improvement. See #99 and also the PyTest portion of `test_qnode.py`. The idea is to remove all loops-inside-tests, and have small, simple, explicit test functions, that test one thing and one thing only. For loops inside tests can also be replaced with `pytest.mark.parametrize` where it makes sense.

**Comment by johannesjmeyer** [comment]
I tried to separate everything as logically sound as possible.

File: tests/test_torch.py [hunk 1]
```diff
@@ -38,116 +35,129 @@
 def expZ(state):
     return np.abs(state[0]) ** 2 - np.abs(state[1]) ** 2
 
+@pytest.fixture()
+def skip_if_no_torch_support(torch_support):
+    if not torch_support:
+        pytest.skip("Skipped, no torch support")
```
Comment (josh146) [inline tests/test_torch.py]: Nice!

File: tests/test_torch.py [hunk 2]
```diff
@@ -38,116 +35,129 @@
...
-        # ...nor observables
-        @qml.qnode(self.dev1, interface='torch')
+        @qml.qnode(qubit_device_1_wire, interface='torch')
         def qf(x):
             return qml.expval(qml.X(0))
 
-        with self.assertRaisesRegex(DeviceError, 'Observable [a-zA-Z]+ not supported on device'):
-            qf(par)
+        with pytest.raises(DeviceError, match='Observable [a-zA-Z]+ not supported on device'):
+            qf(torch.tensor(0.5))
```
Comment (josh146) [inline tests/test_torch.py]: This is fantastic that you managed to separate them like this. A lot of work, but will help make the library a lot more maintainble

File: tests/test_torch.py [hunk 3]
```diff
@@ -165,26 +175,74 @@ def circuit(reused_param, other_param):
...
+        cost_res = cost(*args)
+        cost_res.backward()
+
+        assert np.allclose(cost_res.detach().numpy(), cost_target, atol=tol, rtol=0)
+
+        for i in range(3):
+            assert np.allclose(args[i].grad.detach().numpy(), grad_target[i], atol=tol, rtol=0)
+
+    def test_qnode_array_parameters_1_vector_return(self, qubit_device_1_wire, tol):
+        """Test that QNode can take arrays as input arguments, and that they interact properly with PyTorch."""
```
Comment (josh146) [inline tests/test_torch.py]: docstring is the same as the function above

File: tests/test_torch.py [hunk 4]
```diff
@@ -165,26 +175,74 @@ def circuit(reused_param, other_param):
...
+        cost_res = cost(*args)
+        cost_res.backward()
+
+        assert np.allclose(cost_res.detach().numpy(), cost_target, atol=tol, rtol=0)
+
+        for i in range(3):
+            assert np.allclose(args[i].grad.detach().numpy(), grad_target[i], atol=tol, rtol=0)
+
+    def test_qnode_array_parameters_2_vector_return(self, qubit_device_2_wires, tol):
+        """Test that QNode can take arrays as input arguments, and that they interact properly with PyTorch."""
```
Comment (josh146) [inline tests/test_torch.py]: same as above

File: tests/test_torch.py (hunk 2 above)
Comment (co9olguy) [reply tests/test_torch.py]: lol who wrote them like this originally? :laughing:

File: tests/test_torch.py (hunk 3 above)
Comment (co9olguy) [reply tests/test_torch.py]: Guys, see my comments in PR #266 related to these three tests

File: tests/test_torch.py (hunk 1 above)
Comment (johannesjmeyer) [reply tests/test_torch.py]: I <3 fixtures

File: tests/test_torch.py (hunk 2 above)
Comment (johannesjmeyer) [reply tests/test_torch.py]: I mean it was not an unreasonable amount of work :+1:

File: tests/test_torch.py (hunk 3 above)
Comment (johannesjmeyer) [reply tests/test_torch.py]: Resolved both torch and tfe in the same way

**Comment by johannesjmeyer** [comment]
I also implemented the changes requested by @co9olguy to test_tfe.py here. Should be ready to merge.

**Pull request merged**